docs/user_utilities.rst

```diff
@@ -0,0 +1,154 @@
+
+* Logging
+* Profiler
+
+Logging
+-------
+
+Based on spdlog, Taichi provides a series of APIs for logging. Note that these functions will be called in compile-time but not run-time, if you want to log in run-time, please use ``print()`` instead.
+
+.. Note::
+
+    The current taichi logging API only supports standard output.
+
+.. function:: ti.set_logging_level(level)
+
+    :parameter level: (string) a valid logging level
+
+    - This function is used to set the level of printing logs. Currently, the log levels in taichi are classified from low to high as ``ti.TRACE``, ``ti.DEBUG``, ``ti.INFO``, ``ti.WARN`` and ``ti.ERROR``. These logging levels are essentially strings "trace", "debug", "info", "warn" and "error" respectively.
+
+    - The lower the log level, the more content will be printed out. For example, if you set the print level to ``ti.TRACE``, all logs will be printed; if you set the print level to ``ti.ERROR``, only those log which is generated by ``ti.error()`` will be output.
+
+    - The default logging level is ``ti.INFO``.You can also override the default logging level by setting the environment variable like ``TI_LOG_LEVEL=warn``.
+
+.. function:: ti.info(info)
+
+    :parameter info: (string) logging info
+
+    Print out the input string in green color to stdout, when the logging level is lower or equal to ``ti.INFO``. For example:
+
+.. code-block:: python
+
+    import taichi as ti
+    ti.init(arch = ti.cpu)
+
+    ti.set_logging_level(ti.INFO)
+    var = ti.var(ti.f32, shape = 1)
+
+    @ti.kernel
+    def compute():
+        var[0] = 1.0
+        ti.info(f"set var[0] = 1.0")
+
+    compute()
+
+.. function:: ti.warn(info)
+
+    :parameter info: (string) logging info
+
+    Print out the input string in yellow color to stdout, when the logging level is lower or equal to ``ti.WARN``.
+
+.. function:: ti.debug(info)
+
+    :parameter info: (string) logging info
+
+    Print out the input string in blue color to stdout, when the logging level is lower or equal to ``ti.DEBUG``.
+
+.. function:: ti.trace(info)
+
+    :parameter info: (string) logging info
+
+    Print out the input string in blue color to stdout, only when the logging level is equal to ``ti.TRACE``.
+
+.. function:: ti.error(info)
+
+    :parameter info: (string) logging info
+
+    This function prints out the input string in red color to stdout in any logging level.
+
+    Note that it crashes the Taichi kernel as well as your Taichi program then throw an ``RuntimeError`` out.
+
+.. Warn::
+
+    The usages of ``ti.trace()``, ``ti.debug()`` and ``ti.warn()`` are all similar to ``ti.info()``.But ``ti.error()`` is different from them. When it is called, the program will be crashed immediately.
+
+    In spite of this, the output of ``ti.error()`` is still ``stdout`` but not ``stderr``.
+
+Here is an example:
+
+.. code-block:: python
+
+    import taichi as ti
+
+    ti.init()
+    ti.set_logging_level(ti.ERROR)
```

Comment:
Let's remove this or user would think loglev has to be error to make error work.

Reply:
Great!